openapi2/openapi2.go

```diff
@@ -0,0 +1,310 @@
+		Patch  Operation `json:"patch,omitempty" yaml:"patch,omitempty"`
+		Delete Operation `json:"delete,omitempty" yaml:"delete,omitempty"`
+	}
+
+	// Operation describes a single API operation on a path.
+	Operation struct {
+		OperationID string           `json:"operationId" yaml:"operationId"`
+		Tags        []string         `json:"tags,omitempty" yaml:"tags,omitempty"`
+		Summary     string           `json:"summary,omitempty" yaml:"summary,omitempty"`
+		Description string           `json:"description,omitempty" yaml:"description,omitempty"`
+		Parameters  []interface{}    `json:"parameters,omitempty" yaml:"parameters,omitempty"`
+		Responses   map[int]Response `json:"responses" yaml:"responses"`
+
+		Consumes []string `json:"consumes,omitempty" yaml:"consumes,omitempty"`
+		Produces []string `json:"produces,omitempty" yaml:"produces,omitempty"`
+	}
+
+	// Reference other components in the specification, internally and
+	// externally.
+	Reference struct {
+		Ref string `json:"$ref" yaml:"$ref"`
+	}
+
+	// Response describes a single response from an API Operation.
+	Response struct {
+		Description string
+		Schema      docparse.Schema `json:"schema,omitempty" yaml:"schema,omitempty"`
+	}
+)
+
+// WriteYAML writes w as YAML.
+func WriteYAML(w io.Writer, prog *docparse.Program) error {
+	return write("yaml", w, prog)
+}
+
+// WriteJSON writes to w as JSON.
+func WriteJSON(w io.Writer, prog *docparse.Program) error {
+	return write("json", w, prog)
+}
+
+// WriteJSONIndent writes to w as indented JSON.
+func WriteJSONIndent(w io.Writer, prog *docparse.Program) error {
+	return write("jsonindent", w, prog)
+}
+
+var reParams = regexp.MustCompile(`{\w+}`)
+
+func write(outFormat string, w io.Writer, prog *docparse.Program) error {
+	out := OpenAPI{
+		Swagger: "2.0",
+		Info: Info{
+			Title:   prog.Config.Title,
+			Version: prog.Config.Version,
+			Contact: Contact{
+				Name:  prog.Config.ContactName,
+				Email: prog.Config.ContactEmail,
+				URL:   prog.Config.ContactSite,
+			},
+		},
+		Paths:       map[string]*Path{},
+		Parameters:  map[string]Parameter{},
+		Definitions: map[string]docparse.Schema{},
+
+		Consumes: []string{prog.Config.DefaultRequestCt},
+		Produces: []string{prog.Config.DefaultRequestCt},
+	}
+
+	// Add components.
+	for k, v := range prog.References {
+		if v.Schema == nil {
+			return fmt.Errorf("schema is nil for %v", k)
+		}
+		switch v.Context {
+		case "form", "query", "path":
+			// Nothing, this will be inline in the operation.
+		default:
+			out.Definitions[k] = *v.Schema
+		}
+	}
+
+	// Add endpoints.
+	for _, e := range prog.Endpoints {
+		e.Path = prog.Config.Prefix + e.Path
+
+		op := Operation{
+			Summary:     e.Tagline,
+			Description: e.Info,
+			OperationID: makeID(e),
+			Tags:        e.Tags,
+			Responses:   map[int]Response{},
+		}
+
+		if e.Request.Path != nil {
+			ref := prog.References[e.Request.Path.Reference]
+			for name, p := range ref.Schema.Properties {
```

Comment:
Currently working on converting the tests from OpenAPI3 to 2, and looping over the map there like this means the order is undefined, making tests harder:

```
[~tw/kommentaar](openapi2)% go test -run TestOpenAPI2 -count=10                                                                                                                i:arch:16:14
--- FAIL: TestOpenAPI2 (0.00s)
    --- FAIL: TestOpenAPI2/path-fields (0.00s)
        main_test.go:63: wrong output

                --- expected
                +++ actual
                @@ -14,9 +14,9 @@
                       - tag
                       parameters:
                -      - name: id
                +      - name: companyID
                         in: path
                         required: true
                         type: integer
                -      - name: companyID
                +      - name: id
                         in: path
                         required: true
--- FAIL: TestOpenAPI2 (0.00s)
    --- FAIL: TestOpenAPI2/path-fields (0.00s)
        main_test.go:63: wrong output

                --- expected
                +++ actual
                @@ -14,9 +14,9 @@
                       - tag
                       parameters:
                -      - name: id
                +      - name: companyID
                         in: path
                         required: true
                         type: integer
                -      - name: companyID
                +      - name: id
                         in: path
                         required: true
FAIL
exit status 1
FAIL    github.com/teamwork/kommentaar  0.031s
```

Also means output will change needlessly, making diffs harder and such.

Not sure what the best way to fix this is. We could sort it by name, but does that always make sense? Sometimes you want to group stuff like createdAt/updatedAt/deletedAt, which currently doesn't really work.

Reply:
by type then name perhaps?